gateway-js/src/index.ts

```diff
@@ -425,135 +425,153 @@ export class ApolloGateway implements GraphQLService {
     };
   }
 
-  // Synchronously load a statically configured schema, update class instance's
-  // schema and query planner.
-  private loadStatic(config: StaticGatewayConfig) {
-    let schema: GraphQLSchema;
-    let supergraphSdl: string;
-    try {
-      ({ schema, supergraphSdl } = isLocalConfig(config)
-        ? this.createSchemaFromServiceList(config.localServiceList)
-        : this.createSchemaFromSupergraphSdl(config.supergraphSdl));
-      // TODO(trevor): #580 redundant parse
-      this.parsedSupergraphSdl = parse(supergraphSdl);
-      this.supergraphSdl = supergraphSdl;
-      this.updateWithSchemaAndNotify(schema, supergraphSdl, true);
-    } catch (e) {
-      this.state = { phase: 'failed to load' };
-      throw e;
-    }
-    this.state = { phase: 'loaded' };
+  private getUplinkEndpoints(config: ManagedGatewayConfig) {
+    // 1. If config is set to a `string`, use it
+    // 2. If the env var is set, use that
+    // 3. If config is `undefined`, use the default uplink URLs
+    const rawEndpointsString =
+      process.env.APOLLO_SCHEMA_CONFIG_DELIVERY_ENDPOINT;
+    const envEndpoints = rawEndpointsString?.split(',') ?? null;
+    return config.uplinkEndpoints ??
+      (config.schemaConfigDeliveryEndpoint
+        ? [config.schemaConfigDeliveryEndpoint]
+        : null) ??
+      envEndpoints ?? [
+        'https://uplink.api.apollographql.com/',
+        'https://aws.uplink.api.apollographql.com/',
+      ];
   }
 
-  // Asynchronously load a dynamically configured schema. `this.updateSchema`
-  // is responsible for updating the class instance's schema and query planner.
-  private async loadDynamic(unrefTimer: boolean) {
+  private getIdForSupergraphSdl(supergraphSdl: string) {
+    return createHash('sha256').update(supergraphSdl).digest('hex');
+  }
+
+  private async initializeSupergraphManager<T extends SupergraphManager>(
+    supergraphManager: T,
+  ) {
     try {
-      await this.updateSchema();
+      const result = await supergraphManager.initialize({
+        update: this.externalSupergraphUpdateCallback.bind(this),
+        healthCheck: this.externalSubgraphHealthCheckCallback.bind(this),
+        getDataSource: this.externalGetDataSourceCallback.bind(this),
+      });
+      if (!result?.supergraphSdl) {
+        throw new Error(
+          'Provided `supergraphSdl` function did not return an object containing a `supergraphSdl` property',
+        );
+      }
+      if (result?.cleanup) {
+        if (typeof result.cleanup === 'function') {
+        this.toDispose.push(result.cleanup);
+        } else {
+          this.logger.error(
+            'Provided `supergraphSdl` function returned an invalid `cleanup` property (must be a function)',
+          );
+        }
+      }
+
+      this.externalSupergraphUpdateCallback(result.supergraphSdl);
     } catch (e) {
       this.state = { phase: 'failed to load' };
+      await this.performCleanup();
```

Comment:
I originally wrote this:

> I'd suggest that you do a try/catch/log for exceptions here (ie, if performCleanup throws then we should log its error but allow the original error to still be the thrown one)

But now I see that performCleanup does this itself. Rename to performCleanupAndLogErrors or something?

Reply:
If you prefer, maybe more like performCleanupLoggingErrors or something? I guess the point is that when reading the code I expected the caller to have to be careful about what would happen if performCleanup threw, and it turns out that it doesn't.